### PR TITLE
Error planning tainted resource when it no longer exists

### LINKED ISSUE #27559: invalid planned change after resouce creation failure
While troubleshooting https://github.com/hashicorp/terraform-provider-google/issues/8261 I came upon a bug in terraform core.  In that issue, a provider generates an invalid name for a resource.  The google api rejects the creation.

If I apply again, terraform shows the failed resource as tainted, and exits with an invalid plan error.

### Terraform Version

```
Terraform v0.14.4
+ provider registry.terraform.io/hashicorp/google v3.53.0
+ provider registry.terraform.io/hashicorp/google-beta v3.53.0
```

### Terraform Configuration Files

```terraform
variable "project" {
  type = string
}

provider "google" {
  project = var.project
}

provider "google-beta" {
  project = var.project
}

data "google_compute_network" "network" {
  name = "default"
}

data "google_compute_subnetwork" "subnet" {
  name   = "default"
  region = "us-central1"
}

resource "google_container_cluster" "cluster" {
  provider   = google-beta
  name       = "cluster-1"
  location   = "us-central1"
  network    = data.google_compute_network.network.self_link
  subnetwork = data.google_compute_subnetwork.subnet.self_link

  remove_default_node_pool = true
  node_pool {
    name       = "default-pool"
    node_count = 0
  }

  lifecycle {
    prevent_destroy = true

    ignore_changes = [
      node_pool,
      min_master_version,
    ]
  }
}


resource "google_container_node_pool" "nodes" {
  provider           = google-beta
  name_prefix        = "ca-nonprod-core-"
  location           = "us-central1"
  cluster            = google_container_cluster.cluster.name
  initial_node_count = 1

  node_config {
    machine_type = "n1-standard-4"
    disk_size_gb = 100
    image_type   = "COS_CONTAINERD"

    metadata = {
      disable-legacy-endpoints = "true"
    }

    oauth_scopes = [
      "https://www.googleapis.com/auth/logging.write",
      "https://www.googleapis.com/auth/monitoring",
      "https://www.googleapis.com/auth/devstorage.read_only",
    ]
  }

  lifecycle {
    create_before_destroy = true
  }
}
```

### Debug Output

https://gist.github.com/rvandegrift/b9bc65e5d35c97b80890125e0ca27096

### Expected Behavior

Terraform could continue.  It wants to delete a resource that doesn't exist.

### Actual Behavior

Terraform fails with:
```
Error: Invalid planned change

The plan contains a replace change for google_container_node_pool.nodes, which
doesn't exist yet. This is a bug in Terraform.
```

### References

- https://github.com/hashicorp/terraform-provider-google/issues/8261

### PR DESCRIPTION
**Pull request by jbardin**
When a resource returns an error during apply we "taint" the state for that resource, but save it just in case.
The resource would then be removed during the next refresh if it did not exist, and the plan should result in a Create action rather than a replacement. The tainted state however was compared against `cty.NilVal` rather than checking if it was more generally `Null`, which is the normal case after the resource is removed from the state.

Fixes #27559

Manual 0.14 backport will be required due to changes in 0.15

### CODE REVIEW

File: terraform/node_resource_plan_instance.go [hunk 1]
```diff
@@ -128,11 +128,12 @@ func (n *NodePlannableResourceInstance) managedResourceExecute(ctx EvalContext)
 
 	// Refresh, maybe
 	if !n.skipRefresh {
-		instanceRefreshState, refreshDiags := n.refresh(ctx, instanceRefreshState)
+		s, refreshDiags := n.refresh(ctx, instanceRefreshState)
 		diags = diags.Append(refreshDiags)
 		if diags.HasErrors() {
 			return diags
 		}
+		instanceRefreshState = s
```
Comment (alisdair) [inline terraform/node_resource_plan_instance.go:136]: Is the temporary variable `s` needed here? I don't see the logical difference between this and the prior code, given the shortcut return on error.

**Review by alisdair — APPROVED** [review]
Makes sense to me!

File: terraform/node_resource_plan_instance.go (hunk 1 above)
Comment (jbardin) [reply terraform/node_resource_plan_instance.go:136]: `instanceRefreshState` is passed to `plan` below, but `instanceRefreshState` here is scoped only to this block and shadows the outer variable. The temporary value ensures both that we don't shadow the outer variable, and that the new value is re-assigned.

Comment (alisdair) [reply terraform/node_resource_plan_instance.go:136]: Ah, I see now, thanks for clarifying!

**Pull request merged**